File: next/dweb-browser/src/Platforms/iOS/IpcWeb/MessagePortIpc.cs
using System.Runtime.CompilerServices;
using System.Threading.Tasks.Dataflow;
using DwebBrowser.DWebView;
using DwebBrowser.Helper;
using DwebBrowser.MicroService;
using DwebBrowser.MicroService.Message;
using System.Text.Json;
using Foundation;

#nullable enable

namespace DwebBrowser.IpcWeb;

public class MessagePort
{
    private static ConditionalWeakTable<WebMessagePort, MessagePort> s_wm = new();
    public static MessagePort From(WebMessagePort port) => s_wm.GetValue(port, (port) => new MessagePort(port));

    private WebMessagePort _port = null!;

    private MessagePort(WebMessagePort port)
    {
        _port = port;

        Task.Run(async () =>
        {
            await foreach (var message in MessageChannel.ReceiveAllAsync())
            {
                await (OnWebMessage?.Emit(message)).ForAwait();
            }
        }).Background();

        _port.OnMessage += (message, _) => MessageChannel.SendAsync(message);
    }

    public BufferBlock<WebMessage> MessageChannel = new(new DataflowBlockOptions
    { BoundedCapacity = DataflowBlockOptions.Unbounded });

    public event Signal<WebMessage>? OnWebMessage;

    public Task PostMessage(string data) => _port.PostMessage(WebMessage.From(data, new WebMessagePort[] { _port }));

    private bool _isClosed = false;

    public void Close()
    {
        if (_isClosed)
        {
            MessageChannel.Complete();
            return;
        }

        _isClosed = true;

    }
}

public class MessagePortIpc : Ipc
{
    public MessagePort Port { get; init; }
    private IPC_ROLE _role_type { get; init; }
    public override MicroModuleInfo Remote { get; set; }

    public override string Role { get => Enum.GetName(_role_type)!; }

    public MessagePortIpc(MessagePort port, Ipc.MicroModuleInfo remote, IPC_ROLE role_type)
    {
        Port = port;
        Remote = remote;
        _role_type = role_type;

        var ipc = this;

        Signal<WebMessage> callback = async (message, _) =>
        {
            switch (message.Data)
            {
                case NSString data:
                    switch (MessageToIpcMessage.JsonToIpcMessage((string)data, ipc))
                    {
                        case "close":
                            await Close();
                            break;
                        case "ping":
                            await Port.PostMessage("pong");
                            break;
                        case "pong":
                            Console.WriteLine($"PONG/{ipc}");
                            break;
                        case IpcMessage ipcMessage:
                            Console.WriteLine($"ON-MESSAGE/{ipc} {message}");
                            await _OnMessageEmit(ipcMessage, ipc);
                            break;
                    }
                    break;
                case NSNumber data:
                    Console.WriteLine($"OnWebMessage is number: {data}");
                    break;
                default:
                    throw new Exception($"unknown message: {message}");
            }
        };

        Port.OnWebMessage += callback;
        OnDestory += async (_) => { Port.OnWebMessage -= callback; };
    }

    public override Task _doPostMessageAsync(IpcMessage data)
    {
        string message;
        switch (data)
        {
            case IpcRequest ipcRequest:
                message = ipcRequest.LazyIpcReqMessage.ToJson();
                break;
            case IpcResponse ipcResponse:
                message = ipcResponse.LazyIpcResMessage.ToJson();
                break;
            case IpcStreamData ipcStreamData:
                message = ipcStreamData.ToJson();
                break;
            default:
                message = JsonSerializer.Serialize(data);
                break;
        }

        return Port.PostMessage(message);
    }

    public override async Task DoClose()
    {
        await Port.PostMessage("close");
        Port.Close();
    }
}

File: android/app/src/main/java/info/bagen/dwebbrowser/microService/core/ipc/ipcWeb/MessagePortIpc.kt
package info.bagen.dwebbrowser.microService.core.ipc.ipcWeb

import android.webkit.WebMessage
import android.webkit.WebMessagePort
import info.bagen.dwebbrowser.microService.helper.*
import info.bagen.dwebbrowser.microService.core.ipc.*
import kotlinx.coroutines.CoroutineName
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.launch
import java.util.*

inline fun debugMessagePortIpc(tag: String, msg: Any = "", err: Throwable? = null) =
    printdebugln("message-port-ipc", tag, msg, err)

class MessagePort {
    companion object {
        private val wm = WeakHashMap<WebMessagePort, MessagePort>()
        fun from(port: WebMessagePort): MessagePort = wm.getOrPut(port) { MessagePort(port) }
        val messageScope = CoroutineScope(CoroutineName("webMessage") + ioAsyncExceptionHandler)
    }

    private lateinit var port: WebMessagePort

    private constructor(port: WebMessagePort) {
        this.port = port
    }

    val messageChannel = Channel<WebMessage>(capacity = Channel.UNLIMITED)

    private val _messageSignal by lazy {
        val signal = Signal<WebMessage>()
        messageScope.launch {
            /// 这里为了确保消息的顺序正确性，比如使用channel来一帧一帧地读取数据，不可以直接用 launch 去异步执行 event，这会导致下层解析数据的顺序问题
            /// 并发性需要到消息被解码出来后才能去执行并发。也就是非 IpcStream 类型的数据才可以走并发
            for (event in messageChannel) {
                signal.emit(event)
            }
        }

        port.setWebMessageCallback(object : WebMessagePort.WebMessageCallback() {
            override fun onMessage(port: WebMessagePort, event: WebMessage) {
                messageChannel.trySend(event).getOrThrow()
                /// TODO 尝试告知对方暂停，比如发送 StreamPaused
            }
        })

        signal
    }

    fun onWebMessage(cb: Callback<WebMessage>) = _messageSignal.listen(cb)
    fun postMessage(data: String) = port.postMessage(WebMessage(data))

    private var _isClosed = false
    fun close() {
        if (_isClosed) {
            messageChannel.close()
            return
        }
        _isClosed = true
        port.close()
    }
}

open class MessagePortIpc(
  val port: MessagePort,
  override val remote: MicroModuleInfo,
  private val role_type: IPC_ROLE,
) : info.bagen.dwebbrowser.microService.core.ipc.Ipc() {
    constructor(
        port: WebMessagePort, remote: MicroModuleInfo, role_type: IPC_ROLE
    ) : this(MessagePort.from(port), remote, role_type)

    override val role get() = role_type.role
    override fun toString(): String {
        return super.toString() + "@MessagePortIpc"
    }

    init {
        val ipc = this;
        val callback = port.onWebMessage { event ->
            when (val message = jsonToIpcMessage(event.data, ipc)) {
                "close" -> close()
                "ping" -> port.postMessage("pong")
                "pong" -> debugMessagePortIpc("PONG/$ipc")
                is IpcMessage -> {
                    debugMessagePortIpc("ON-MESSAGE/$ipc", message)
                    _messageSignal.emit(
                        IpcMessageArgs(message, ipc)
                    )
                }
                else -> throw Exception("unknown message: $message")
            }
        }
        onDestroy(callback)

    }

    override suspend fun _doPostMessage(data: IpcMessage) {
        val message = when (data) {
            is IpcRequest -> gson.toJson(data.ipcReqMessage)
            is IpcResponse -> gson.toJson(data.ipcResMessage)
            is IpcStreamData -> gson.toJson(data)
            else -> gson.toJson(data)
        }
        this.port.postMessage(message)
    }

    override suspend fun _doClose() {
        this.port.postMessage("close")
        this.port.close()
    }
}

